Nodes/modules/puzzle_coap/puzzle_coap.c

```diff
@@ -0,0 +1,282 @@
+        return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+}
+
+
+void puzzle_init(const puzzle_t *puzzle)
+{
+    assert(puzzle);
+    assert(puzzle->name);
+    assert(puzzle->resource_dir_uri);
+    assert(puzzle->get_solved_handler);
+    assert(puzzle->get_ready_handler);
+    assert(puzzle->set_ready_handler);
+    _puzzle_info = puzzle;
+
+    sock_udp_ep_t remote;
+    uri_parser_result_t uri_result;
+    assert(uri_parser_is_absolute_string(puzzle->resource_dir_uri));
+    assert(uri_parser_process_string(&uri_result, puzzle->resource_dir_uri) == 0);
+    assert(_make_sock_ep(&remote, &uri_result) == 0);
+
+    gcoap_register_listener(&_default_listener);
+
+    /*
+    puts("Registering with RD now, this may take a short while...");
+    if (cord_ep_register(&remote, NULL) != CORD_EP_OK) {
+        puts("error: registration failed");
+        return;
+    } else {
+        puts("registration successfull");
+    }
+    */
+}
+
+
+void puzzle_update(void)
+{
+    size_t len = 0;
+    uint8_t buf[CONFIG_GCOAP_PDU_BUF_SIZE];
+    coap_pkt_t pdu;
+
+    switch (gcoap_obs_init(&pdu, buf, sizeof(buf), &_resources[0])) {
+        case GCOAP_OBS_INIT_OK:
+            len = _build_cbor_coap_paket(&pdu, buf, sizeof(buf));
+            if(gcoap_obs_send(buf, len, &_resources[0]) == 0) {
+                puts("Notification failed to send");
+            }
+            break;
+        case GCOAP_OBS_INIT_ERR:
+            puts("Notification of the observers failed");
+            break;
+        case GCOAP_OBS_INIT_UNUSED:
+            break;
+    }
+}
+
+/*
+ * Server callback for /node/info. Accepts only GET.
+ *
+ * GET: Returns the puzzle status in CBOR
+ */
+static ssize_t _puzzle_info_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx)
+{
+    (void)ctx;
+    /* since the init functions asserts all other members, 
+     * its safe to just assert the base pointer */
+    assert(_puzzle_info);
+
+    /* initialize a new coap response */
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+
+    return _build_cbor_coap_paket(pdu, buf, len);
+
+}
+
+/*
+ * Server callback for /node/ready. Accepts only PUT.
+ * Sets the puzzle into the ready state
+ * PUT: Returns the puzzle status in CBOR
+ */
+static ssize_t _puzzle_ready_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx)
+{
+    (void)ctx;
+    /* since the init functions asserts all other members, 
+     * its safe to just assert the base pointer */
+    assert(_puzzle_info);
+
+    unsigned method = coap_method2flag(coap_get_code_detail(pdu));
+
+    switch (method) {
+        case COAP_PUT:
+            _puzzle_info->set_ready_handler(false);
+            break;
+        default:
+            /* we don't care about anything else */
+            return gcoap_response(pdu, buf, len, COAP_CODE_BAD_REQUEST);
+    }
+
+
+    /* initialize a new coap response */
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
```

Comment:
https://datatracker.ietf.org/doc/html/rfc7252#section-5.8.3
In general we would return the current representation of the modified resource. Maybe just the puzzle state part. Also, the return code should be CHANGED.

Reply:
> In general we would return the current representation of the modified resource.

Yes. And that's already what's going on in the code.
```C
return _build_cbor_coap_packet(pdu, buf, len);
```

> Maybe just the puzzle state part. 
Hm yes maybe. 🤔